### DISCUSSION TITLE
Aggregator weights by object ID

### DISCUSSION THREAD

**Discussion by Naarlack** [anitagraser/movingpandas#364] [Q&A] [answered]
I have a bunch of GPS data I have collected from multiple cyclists. I would like to show a map of where they have ridden, but would also like to preserve their individual privacy. I'm a new MovingPandas user but I have managed to successfully get their data into a TrajectoryCollection with "traj_id_col" label for each trajectory and "obj_id_col" label for each rider.

I then followed the tutorial on using the TrajectoryCollectionAggregator to plot the flows. The "weight" column in that flows data allows me to control for the number of trajectories but gives no information on the number of objects. Is there any way I could achieve that, or some kind of work around, so that I only show the flow lines that more than one rider has traversed?

**Comment by raybellwaves** [comment]
I don't suppose you could create some dummy data and provide a MCVE (https://blog.dask.org/2018/02/28/minimal-bug-reports) to show what you've done so we can have a go at tweaking it to what you want

**Comment by Naarlack** [comment]
Sure, I'll put something together and get back to you. Thanks for taking a look.

**Reply by Naarlack** [reply]
Sorry about the longform formatting of the data points. Ruff was not playing nice 😒

The dummy data shows three journeys by two riders. One rider does the same route twice, while the other does a different route that has some overlap with the other route.

I have hard-coded a 'riders' column to the aggregator flows output as an example of what I would like to achieve, and plotted for further clarity.

Thanks again 😊

```python
import pandas as pd
import geopandas as gpd
import movingpandas as mpd
from shapely.geometry import Point
from datetime import datetime, timedelta
import matplotlib.pyplot as plt

df = pd.DataFrame(
    [
        {
            "geometry": Point(0, 0),
            "t": datetime(2024, 1, 1, 12, 0, 0),
            "rider_id": "001",
            "journey_id": 1,
        },
        {
            "geometry": Point(6, 0),
            "t": datetime(2024, 1, 1, 12, 0, 1),
            "rider_id": "001",
            "journey_id": 1,
        },
        {
            "geometry": Point(6, 6),
            "t": datetime(2024, 1, 1, 12, 0, 2),
            "rider_id": "001",
            "journey_id": 1,
        },
        {
            "geometry": Point(12, 12),
            "t": datetime(2024, 1, 1, 12, 0, 3),
            "rider_id": "001",
            "journey_id": 1,
        },
        {
            "geometry": Point(0, 0),
            "t": datetime(2024, 1, 2, 12, 0, 0),
            "rider_id": "001",
            "journey_id": 2,
        },
        {
            "geometry": Point(6, 0),
            "t": datetime(2024, 1, 2, 12, 0, 1),
            "rider_id": "001",
            "journey_id": 2,
        },
        {
            "geometry": Point(6, 6),
            "t": datetime(2024, 1, 2, 12, 0, 2),
            "rider_id": "001",
            "journey_id": 2,
        },
        {
            "geometry": Point(12, 12),
            "t": datetime(2024, 1, 2, 12, 0, 3),
            "rider_id": "001",
            "journey_id": 2,
        },
        {
            "geometry": Point(12, 0),
            "t": datetime(2024, 1, 3, 12, 0, 0),
            "rider_id": "002",
            "journey_id": 3,
        },
        {
            "geometry": Point(6, 0),
            "t": datetime(2024, 1, 3, 12, 0, 1),
            "rider_id": "002",
            "journey_id": 3,
        },
        {
            "geometry": Point(6, 6),
            "t": datetime(2024, 1, 3, 12, 0, 2),
            "rider_id": "002",
            "journey_id": 3,
        },
        {
            "geometry": Point(12, 12),
            "t": datetime(2024, 1, 3, 12, 0, 3),
            "rider_id": "002",
            "journey_id": 3,
        },
    ]
).set_index("t")

gdf = gpd.GeoDataFrame(df, crs=31256)

traj_collection = mpd.TrajectoryCollection(
    gdf, traj_id_col="journey_id", obj_id_col="rider_id", t="t", x="X", y="Y"
)

aggregator = mpd.TrajectoryCollectionAggregator(
    traj_collection, max_distance=5, min_distance=0.5, min_stop_duration=timedelta(seconds=0.5)
)

flows = aggregator.get_flows_gdf()

# This is the information I would like to somehow get from the aggregator
flows["riders"] = [1, 2, 2, 1]

fig = plt.figure(figsize=(10, 5))
axL = plt.subplot(121)
axR = plt.subplot(122)

# Shows number of journeys from any rider - including multiple from the same rider
flows.plot(ax=axL, color="green", linewidth=flows["weight"], alpha=0.6)
# Shows only the journeys where more than one rider has travelled [this is what I want]
flows[flows["riders"] > 1].plot(ax=axR, color="blue", linewidth=flows["weight"], alpha=0.6)

plt.show()
```

**Reply by raybellwaves** [reply, marked as answer]
I think what you would like is a obj_id_col_count arg (bool) to `get_flows_gdf`.

Here is a quick and dirty post-processed version. It's likely this information is in the code and just not being returned.

```
flows['start_point'] = flows['geometry'].apply(lambda geom: Point(geom.coords[0]))
flows['end_point'] = flows['geometry'].apply(lambda geom: Point(geom.coords[-1]))
counts = []
for ix, flow in flows.iterrows():
    start_point = flow['start_point']
    end_point = flow['end_point']
    count = 0
    for rider_id in gdf["rider_id"].unique():
        trajs = traj_collection.filter("rider_id", rider_id)
        for traj in trajs:
            traj_geom = traj.to_linestring()
            if start_point.within(traj_geom.buffer(0.0001)) and end_point.within(traj_geom.buffer(0.0001)):
                count += 1
                break
    counts.append(count)
```

If you have time I encourage you to see how easy/hard it would be to add to the source code

**Reply by Naarlack** [reply]
Ahhh, this is nice. Thanks! 😁

I assume I would need to be quite careful with how I set buffer sizes and such when using it beyond the current toy example?

I have been looking at the source code, and trying to figure out where it could go. I'm thinking somewhere in `_SequenceGenerator`, but I'm not quite sure how to build upon the current `Counter` object implementation. In addition, I'm not sure what the output should be. Would adding an "object_weight" column (in addition to the current "weight" column) make sense?

I'm quite new to collaborating through Github, but would love to give it a go (if you're willing to put up with my inevitable mistakes 😅).

**Reply by anitagraser** [reply]
This would be a great addition if you get around to coding it
